pyccel/ast/numpyext.py

```diff
@@ -90,12 +90,15 @@
     'pi': Constant('real', 'pi', value=numpy.pi),
 }
 
+class PyccelAllocatableObject(PyccelAstNode):
+    pass
```

Comment:
What does this class represent? Could you add a docstring to it?

By the way, it is a little strange to use a name containing "Object" for a class...

Reply:
This is related to issue #307. The class represents objects which are allocatable. These objects then inherit from this. This includes `Full`, `Linspace`, `Diag`, `Where`, etc
I'm not sure what exactly to put in a docstring given that my explanation of `AllocatableObject` is that it is an object that is allocatable ^^'.

I can change the name if you prefer. Should it just be called `PyccelAllocatable` ?

Reply:
I see. Maybe more explicitly, I would say in the docstring that this class represents Python constructs corresponding to a memory allocation on the heap (unless the `@stack_array` decorator is used), and which we only accept as the right-hand side of an assignment.

Since this is only relevant to arrays at the moment, maybe this class should be called `PyccelArray` and be moved to a different module. And maybe we need to distinguish between a `PyccelStackArray` and a `PyccelHeapArray`, and return the correct one in the `__new__` constructor.

Sorry for all the "maybe"! @saidctb What do you think?

Reply:
Actually, for this PR it should be enough to change the name of this class to `NumpyNewArray` and leave it where it is. We can discuss later about further generalizations.